api/client.js

```diff
@@ -120,56 +120,62 @@ function connect(url, options={}, callback=null) {
       level;
-    - wsclass (default=W3C browser native WebSocket): the WebSocket class to use
-      for opening the connection and sending/receiving messages. Server side,
-      require('websocket').w3cwebsocket can be used safely, as it implements the
-      W3C browser native WebSocket API;
+    - wsclass (default=W3C browser native WebSocket): the WebSocket class to
+      use for opening the connection and sending/receiving messages. Server
+      side, require('websocket').w3cwebsocket can be used safely, as it
+      implements the W3C browser native WebSocket API;
     - bakery (default: null): the bakery client to use when macaroon discharges
       are required, in the case an external user is used to connect to Juju;
       see <https://www.npmjs.com/package/macaroon-bakery>;
     - closeCallback: a callback to be called with the exit code when the
       connection is closed.
   @param {Function} callback Called when the login process completes, the
-    callback receives an error, a connection object and a logout function that
-    can be used to clos ethe connection. If there are no errors, the connection
-    can be used to send/receive messages to and from the Juju controller or
-    model, and to get access to the available facades (through conn.facades).
-    See the docstring for the Connection class for information on how to use
-    the connection instance. Redirection errors are automatically handled by
-    this function, so any error is a real connection problem.
+    callback receives an error, or a {conn, logout} result including a
+    connection object and a logout function that can be used to close the
+    connection. If there are no errors, the connection can be used to
+    send/receive messages to and from the Juju controller or model, and to get
+    access to the available facades (through conn.facades). See the docstring
+    for the Connection class for information on how to use the connection
+    instance. Redirection errors are automatically handled by this function, so
+    any error is a real connection problem.
+  @return {Promise} This promise will be rejected if there is an error
+    connecting, or resolved with a new {conn, logout} object. Note that the
+    promise will not be resolved or rejected if a callback is provided.
 */
 function connectAndLogin(url, credentials, options, callback) {
-  // Connect to Juju.
-  connect(url, options, (err, juju) => {
-    if (err) {
-      callback(err, null, null);
-      return;
-    }
-    // Authenticate.
-    juju.login(credentials, (err, conn) => {
-      if (!err) {
-        callback(null, conn, juju.logout.bind(juju));
-        return;
-      }
-      if (!juju.isRedirectionError(err)) {
-        callback(err, null, null);
+  return new Promise((resolve, reject) => {
+    const handler = createAsyncHandler(callback, resolve, reject);
+    // Connect to Juju.
+    connect(url, options, (err, juju) => {
```

Comment:
We should standardize on using either the callback or async signature when writing the internal convenience methods.

I believe that this is the first method so you can choose if you want to instead use the async versions of these methods :)

Reply:
In this case it felt easier to me using callbacks. Maybe because it required no code changes. We can discuss what to use.